Source/WTF/wtf/RunLoop.h

```diff
@@ -114,11 +118,27 @@ class WTF_CAPABILITY("is current") RunLoop final : public GuaranteedSerialFuncti
 
     WTF_EXPORT_PRIVATE void threadWillExit();
 
+    enum class Activity : uint8_t {
+        BeforeWaiting   = 1 << 0,
+        Entry           = 1 << 1,
+        Exit            = 1 << 2,
+        AfterWaiting    = 1 << 3
+    };
+
 #if USE(GLIB_EVENT_LOOP)
     WTF_EXPORT_PRIVATE GMainContext* mainContext() const { return m_mainContext.get(); }
-    enum class Event { WillDispatch, DidDispatch };
-    using Observer = WTF::Observer<void(Event, const String&)>;
-    WTF_EXPORT_PRIVATE void observe(const Observer&);
+
+    // Event observers (only used for PageTimelineAgent)
+    enum class Event : bool {
+        WillDispatch,
+        DidDispatch
+    };
+
+    using EventObserver = Observer<void(Event, const String&)>;
+    WTF_EXPORT_PRIVATE void observeEvent(const EventObserver&);
+
+    WTF_EXPORT_PRIVATE void observeActivity(const RefPtr<ActivityObserver>&);
+    WTF_EXPORT_PRIVATE void unobserveActivity(const RefPtr<ActivityObserver>&);
```

Comment:
I see that scheduling/unscheduling events and stopping the RunLoop is all thread-safe. Probably the observers should be too? But it is not, because `m_activityObservers` and `m_eventObservers` is not locked.